core/modules/wsched/ChunkTaskList.h

```diff
@@ -0,0 +1,160 @@
+
+// System headers
+#include <algorithm>
+#include <list>
+#include <map>
+#include <mutex>
+
+// Qserv headers
+#include "memman/MemMan.h"
+#include "wbase/Task.h"
+#include "wsched/ChunkTaskCollection.h"
+
+namespace lsst {
+namespace qserv {
+namespace wsched {
+
+
+/// A class to store Tasks for a specific chunk.
+//  Tasks are normally placed on _activeTasks, but will be added
+//  to _pendingTasks when this is the active chunk.
+class ChunkTasks {
+public:
+    using Ptr = std::shared_ptr<ChunkTasks>;
+    enum class ReadyState {READY, NOT_READY, NO_RESOURCES};
+
+    ChunkTasks(int chunkId, memman::MemMan::Ptr const& memMan) : _chunkId{chunkId}, _memMan{memMan} {}
+    ChunkTasks() = delete;
+    ChunkTasks(ChunkTasks const&) = delete;
+    ChunkTasks& operator=(ChunkTasks const&) = delete;
+
+    bool empty() const;
+    void queTask(wbase::Task::Ptr const& task);
+    wbase::Task::Ptr getTask(bool useFlexibleLock);
+    ReadyState ready(bool useFlexibleLock);
+    void taskComplete(wbase::Task::Ptr const& task);
+
+    void movePendingToActive(); ///< Move all pending Tasks to _activeTasks.
+    bool readyToAdvance(); ///< @return true if active Tasks for this chunk are done.
+    void setActive(bool active=true); ///< Flag current requests so new requests will be pending.
+    bool setResourceStarved(bool starved); ///< hook for tracking starvation.
+    std::size_t size() const { return _activeTasks.size() + _pendingTasks.size(); }
+    int getChunkId() { return _chunkId; }
+
+    /// Class that keeps the slowest tables at the front of the heap.
+    class SlowTableHeap {
+    public:
+        // Using a greater than comparison function results in a minimum value heap.
+        static bool compareFunc(wbase::Task::Ptr const& x, wbase::Task::Ptr const& y) {
+            if(!x || !y) { return false; }
+            // compare scanInfo (slower scans first)
+            int siComp = x->getScanInfo().compareTables(y->getScanInfo());
+            return siComp < 0;
+        };
+        void push(wbase::Task::Ptr const& task);
+        wbase::Task::Ptr pop();
+        wbase::Task::Ptr top() {
+            if (_tasks.empty()) return nullptr;
+            return _tasks.front();
+        }
+        bool empty() const { return _tasks.empty(); }
+        size_t size() const { return _tasks.size(); }
+        void heapify() {
+            std::make_heap(_tasks.begin(), _tasks.end(), compareFunc);
+        }
+
+        std::vector<wbase::Task::Ptr> _tasks;
+    };
+
+private:
+    int _chunkId;          ///< Chunk Id for all Tasks in this instance.
+    bool _active{false};   ///< True when this is the active chunk.
+    bool _resourceStarved; ///< True when advancement is prevented by lack of memory.
+    wbase::Task::Ptr              _readyTask{nullptr}; ///< Task that is ready to run with memory reserved.
+    SlowTableHeap                 _activeTasks;        ///< All Tasks must be put on this before they can run.
```

Comment:
I've been basing "in flight" on Daniel's past code comments. "in flight" basically means running, but to avoid race conditions, a Task is considered "in flight" as soon as getTask() pulls it off the heap.